clippy_lints/src/nonstandard_macro_braces.rs

```diff
@@ -0,0 +1,246 @@
+            emit_help(cx, snip, braces, name, item.span);
+            n = Some(name.to_owned());
+        }
+        if let Some(name) = n {
+            self.done.insert(name);
+        }
+    }
+
+    fn check_stmt(&mut self, cx: &EarlyContext<'_>, stmt: &ast::Stmt) {
+        let mut n = None;
+        if let Some((name, braces, snip)) = is_offending_macro(cx, stmt.span, self) {
+            emit_help(cx, snip, braces, name, stmt.span);
+            n = Some(name.to_owned());
+        }
+        if let Some(name) = n {
+            self.done.insert(name);
+        }
+    }
+
+    fn check_expr(&mut self, cx: &EarlyContext<'_>, expr: &ast::Expr) {
+        let mut n = None;
+        if let Some((name, braces, snip)) = is_offending_macro(cx, expr.span, self) {
+            emit_help(cx, snip, braces, name, expr.span);
+            n = Some(name.to_owned());
+        }
+        if let Some(name) = n {
+            self.done.insert(name);
+        }
+    }
+}
+
+fn is_offending_macro<'a>(cx: &EarlyContext<'_>, span: Span, this: &'a MacroBraces) -> Option<MacroInfo<'a>> {
+    if_chain! {
+        if in_macro(span);
+        if let Some(MacroMatcher { name, braces }) = find_matching_macro(span, &this.macro_braces);
+        if let Some(snip) = snippet_opt(cx, span.ctxt().outer_expn_data().call_site);
+        let c = snip.replace(" ", ""); // make formatting consistent
+        if !c.starts_with(&format!("{}!{}", name, braces.0));
+        if !this.done.contains(name.as_str());
+        then {
+            Some((name, braces, snip))
+        } else {
+            None
+        }
+    }
+}
+
+fn emit_help(cx: &EarlyContext<'_>, snip: String, braces: &(String, String), name: &str, span: Span) {
+    let mut help = snip;
+    for b in BRACES.iter().filter(|b| b.0 != braces.0) {
+        help = help.replace(b.0, &braces.0).replace(b.1, &braces.1);
+        // Only `{` traditionally has space before the brace
+        if braces.0 != "{" && help.contains(&format!(" {}", braces.0)) {
+            help = help.replace(&format!(" {}", braces.0), &braces.0);
+        }
+    }
+    span_lint_and_help(
+        cx,
+        NONSTANDARD_MACRO_BRACES,
+        span,
+        &format!("use of irregular braces for `{}!` macro", name),
+        Some(span.ctxt().outer_expn_data().call_site),
+        &format!("consider writing `{}`", help),
+    );
+}
+
+fn find_matching_macro(span: Span, braces: &[MacroMatcher]) -> Option<&MacroMatcher> {
+    braces.iter().find(|mac| is_direct_expn_of(span, &mac.name).is_some())
+}
+
+macro_rules! macro_matcher {
+    (name: $name:expr, braces: ($open:expr, $close:expr) $(,)?) => {
+        MacroMatcher {
+            name: $name.to_owned(),
+            braces: ($open.to_owned(), $close.to_owned()),
+        }
+    };
+}
+pub(crate) use macro_matcher;
+
+fn macro_braces(conf: FxHashSet<MacroMatcher>) -> Vec<MacroMatcher> {
+    let mut braces = vec![
```

Comment:
I think `braces` should be a Set instead of a `Vec`. The way it is now does not allow `conf` to determinisitically override settings.